File: LowLevelApi/NGPF/Python/Statistics/Sample_Statistics_Operations.py
################################################################################
# Version 1.0    $Revision: 1 $
# $Author: Mario Dicu $
#
#    Copyright © 1997 - 2015 by IXIA
#    All Rights Reserved.
#
################################################################################

################################################################################
#                                                                              #
#                                LEGAL  NOTICE:                                #
#                                ==============                                #
# The following code and documentation (hereinafter "the script") is an        #
# example script for demonstration purposes only.                              #
# The script is not a standard commercial product offered by Ixia and have     #
# been developed and is being provided for use only as indicated herein. The   #
# script [and all modifications, enhancements and updates thereto (whether     #
# made by Ixia and/or by the user and/or by a third party)] shall at all times #
# remain the property of Ixia.                                                 #
#                                                                              #
# Ixia does not warrant (i) that the functions contained in the script will    #
# meet the user's requirements or (ii) that the script will be without         #
# omissions or error-free.                                                     #
# THE SCRIPT IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, AND IXIA        #
# DISCLAIMS ALL WARRANTIES, EXPRESS, IMPLIED, STATUTORY OR OTHERWISE,          #
# INCLUDING BUT NOT LIMITED TO ANY WARRANTY OF MERCHANTABILITY AND FITNESS FOR #
# A PARTICULAR PURPOSE OR OF NON-INFRINGEMENT.                                 #
# THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SCRIPT  IS WITH THE #
# USER.                                                                        #
# IN NO EVENT SHALL IXIA BE LIABLE FOR ANY DAMAGES RESULTING FROM OR ARISING   #
# OUT OF THE USE OF, OR THE INABILITY TO USE THE SCRIPT OR ANY PART THEREOF,   #
# INCLUDING BUT NOT LIMITED TO ANY LOST PROFITS, LOST BUSINESS, LOST OR        #
# DAMAGED DATA OR SOFTWARE OR ANY INDIRECT, INCIDENTAL, PUNITIVE OR            #
# CONSEQUENTIAL DAMAGES, EVEN IF IXIA HAS BEEN ADVISED OF THE POSSIBILITY OF   #
# SUCH DAMAGES IN ADVANCE.                                                     #
# Ixia will not be required to provide any software maintenance or support     #
# services of any kind (e.g., any error corrections) in connection with the    #
# script or any part thereof. The user acknowledges that although Ixia may     #
# from time to time and in its sole discretion provide maintenance or support  #
# services for the script, any such services are subject to the warranty and   #
# damages limitations set forth herein and will not obligate Ixia to provide   #
# any additional maintenance or support services.                              #
#                                                                              #
################################################################################

################################################################################
#                                                                              #
# Description:                                                                 #
#    This sample configures 10 IPv4 sessions on each of the two ports,         # 
#    adds a traffic Item that uses IPv4 endpoints, sends traffic,              #
#    using statistics, the performs the following actions:                     #
#    - enable/disable CSV Logging                                              #
#    - Add Formula Column to view                                              #
#    - Edit the Formula Column added to view                                   #
#    - Take a Snapshot CSV for view Flow Statistics                            #
#    - Check the Tx Frames = Rx Frames for each IPv4 source address            #
#                                                                              #
################################################################################

if 'py' not in dir():
    class TestFailedError(Exception): pass
    class Py: pass
    py = Py()
    py.ports = [('10.205.15.62', 3, 5), ('10.205.15.62', 3, 6)]
    py.ixTclServer = '10.205.15.224'
    py.ixTclPort = 8009

################################################################################
# Import the IxNet library
################################################################################
from IxNetwork import IxNet
import time
ixNet = IxNet()

################################################################################
# Connect to IxNet client
################################################################################

ixNet.connect(py.ixTclServer, '-port', py.ixTclPort, '-version', '7.40')

################################################################################
# Cleaning up IxNetwork
################################################################################
print "Cleaning up IxNetwork..."
ixNet.execute('newConfig')

################################################################################
# Adding ports to configuration
################################################################################
print "Adding ports to configuration"
root = ixNet.getRoot()
ixNet.add(root, 'vport')
ixNet.add(root, 'vport')
ixNet.commit()
vPorts = ixNet.getList(root, 'vport')
vport1 = vPorts[0]
vport2 = vPorts[1]

################################################################################
# Adding IPv4 endpoints to configuration
################################################################################
print "Add topologies"
ixNet.add(root, 'topology')
ixNet.add(root, 'topology')
ixNet.commit()

topo1 = ixNet.getList(root, 'topology')[0]
topo2 = ixNet.getList(root, 'topology')[1]

print "Add ports to topologies"
ixNet.setAttribute(topo1, '-vports', vport1)
ixNet.setAttribute(topo2, '-vports', vport2)
ixNet.commit()

print "Add device groups to topologies"
ixNet.add(topo1, 'deviceGroup')
ixNet.add(topo2, 'deviceGroup')
ixNet.commit()

dg1 = ixNet.getList(topo1, 'deviceGroup')[0]
dg2 = ixNet.getList(topo2, 'deviceGroup')[0]

print "Add ethernet stacks to device groups"
ixNet.add(dg1, 'ethernet')
ixNet.add(dg2, 'ethernet')
ixNet.commit()

mac1 = ixNet.getList(dg1, 'ethernet')[0]
mac2 = ixNet.getList(dg2, 'ethernet')[0]

print "Add ipv4 stacks to ethernets"
ixNet.add(mac1, 'ipv4')
ixNet.add(mac2, 'ipv4')
ixNet.commit()

ipv4_1 = ixNet.getList(mac1, 'ipv4')[0]
ipv4_2 = ixNet.getList(mac2, 'ipv4')[0]

print "Setting multi values for ipv4 addresses"
ixNet.setMultiAttribute(ixNet.getAttribute(ipv4_1, '-address') + '/counter', '-start', '22.1.1.1', '-step', '0.0.1.0')
ixNet.setMultiAttribute(ixNet.getAttribute(ipv4_1, '-gatewayIp') + '/counter', '-start', '22.1.1.2', '-step', '0.0.1.0')
ixNet.setMultiAttribute(ixNet.getAttribute(ipv4_1, '-resolveGateway') + '/singleValue', '-value', 'true')
ixNet.setMultiAttribute(ixNet.getAttribute(ipv4_2, '-address') + '/counter', '-start', '22.1.1.2', '-step', '0.0.1.0')
ixNet.setMultiAttribute(ixNet.getAttribute(ipv4_2, '-gatewayIp') + '/counter', '-start', '22.1.1.1', '-step', '0.0.1.0')
ixNet.setMultiAttribute(ixNet.getAttribute(ipv4_2, '-resolveGateway') + '/singleValue', '-value', 'true')
ixNet.commit()

################################################################################
# Create Traffic for IPv4
################################################################################
print ''
print "Creating Traffic for IPv4"

ixNet.add(ixNet.getRoot() + '/traffic', 'trafficItem')
ixNet.commit()
ti1 = ixNet.getList(ixNet.getRoot() + '/traffic', 'trafficItem')[0]
ixNet.setMultiAttribute( ti1,
        '-name'                 ,'Traffic IPv4',
        '-trafficType'          ,'ipv4',
        '-allowSelfDestined'    ,False,
        '-trafficItemType'      ,'l2L3',
        '-mergeDestinations'    ,True,
        '-egressEnabled'        ,False,
        '-srcDestMesh'          ,'manyToMany',
        '-enabled'              ,True,
        '-routeMesh'            ,'fullMesh',
        '-transmitMode'         ,'interleaved',
        '-biDirectional'        ,True,
        '-hostsPerNetwork'      ,1)
ixNet.commit()
ixNet.setAttribute(ti1, '-trafficType', 'ipv4')
ixNet.commit()
ixNet.add(ti1, 'endpointSet',
        '-sources',             ipv4_1,
        '-destinations',        ipv4_2,
        '-name',                'ep-set1',
        '-sourceFilter',        '',
        '-destinationFilter',   '')
ixNet.commit()
ixNet.setMultiAttribute(ti1 + "/configElement:1/frameSize",
        '-type',        'fixed',
        '-fixedSize',   128)
ixNet.setMultiAttribute(ti1 + "/configElement:1/frameRate",
        '-type',        'percentLineRate',
        '-rate',        10)
ixNet.setMultiAttribute(ti1 + "/configElement:1/transmissionControl",
        '-duration'               ,1,
        '-iterationCount'         ,1,
        '-startDelayUnits'        ,'bytes',
        '-minGapBytes'            ,12,
        '-frameCount'             ,10000,
        '-type'                   ,'fixedFrameCount',
        '-interBurstGapUnits'     ,'nanoseconds',
        '-interBurstGap'          , 0,
        '-enableInterBurstGap'    ,False,
        '-interStreamGap'         ,0,
        '-repeatBurst'            ,1,
        '-enableInterStreamGap'   ,False,
        '-startDelay'             ,0,
        '-burstPacketCount'       ,1,)
ixNet.setMultiAttribute(ti1 + "/tracking", '-trackBy', ['ipv4SourceIp0'])
ixNet.commit()

################################################################################
# Assign ports 
################################################################################
vports = ixNet.getList(ixNet.getRoot(), 'vport')
print "Assigning ports to " + str(vports) + " ..."
assignPorts = ixNet.execute('assignPorts', py.ports, [], ixNet.getList("/","vport"), True)
if assignPorts != vports:
    raise TestFailedError("FAILED assigning ports. Got %s" %assignPorts)
else:
    print("PASSED assigning ports. Got %s" %assignPorts)

################################################################################
# Start All Protocols
################################################################################
print "Starting All Protocols"
ixNet.execute('startAllProtocols')
print "Sleep 30sec for protocols to start"
time.sleep(30)

################################################################################
# Generate, apply, start traffic
################################################################################
ixNet.execute('generate', ti1)
ixNet.execute('apply', '/traffic')
ixNet.execute('start', '/traffic')
print "Sleep 30sec to send all traffic"
time.sleep(30)

print "#########################"
print "## Statistics Samples ##"
print "#########################"
print ""

################################################################################
# Define function to get the view object using the view name
################################################################################
def getViewObject(ixNet, viewName):
    views = ixNet.getList('/statistics', 'view')
    viewObj = ''
    editedViewName = '::ixNet::OBJ-/statistics/view:\"' + viewName + '\"'
    for view in views:
        if editedViewName == view:
             viewObj = view
             break
    return viewObj

################################################################################
# Define function to get the values for the statistics in the view
################################################################################
def getValuesForStatInView(ixNet, viewName, statName):
    print "- get the stats for %s in view %s" % ( statName, viewName )
    views = ixNet.getList('/statistics', 'view')
    viewObj = getViewObject(ixNet, viewName)
    returned_values = ixNet.execute('getColumnValues', viewObj, statName)
    return returned_values

################################################################################
# Define function to get all the statistics in the view
################################################################################
def getAllStatsInView(ixNet, viewName):
    print "- get the stats in view %s" % viewName
    mview = getViewObject(ixNet, viewName)
    mpage = ixNet.getList(mview, 'page')[0]
    mrowvalues = ixNet.getAttribute(mpage, '-rowValues')
    return mrowvalues

################################################################################
# Define function to create a Snapshot CSV
################################################################################
def takeViewCSVSnapshot(ixNet, viewName, csvPath="c:\\Regression\\Snapshot CSVs", csvType="currentPage"):
    print "- take Snapshot CSV"
    SnapSettingList = [ 'Snapshot.View.Csv.Location: "' + csvPath + '"',
                        'Snapshot.View.Csv.GeneratingMode: "kOverwriteCSVFile"',
                        'Snapshot.Settings.Name: ' + viewName, 
                        'Snapshot.View.Contents: ' + csvType ]
    ixNet.execute('TakeViewCSVSnapshot',str('"' + viewName + '"'),SnapSettingList)
    print "- snapshot CSV complete"

################################################################################
# Define function to Enable CSV Logging
################################################################################
def setEnableCsvLogging(ixNet, state=False):
    print "- set enableCsvLogging to: %s" % state
    ixNet.setAttribute('/statistics', '-enableCsvLogging', state)
    ixNet.commit()

################################################################################
# Define function to add formula column
################################################################################
def addFormulaColumn(ixNet, viewName, columnName, formula):
    print "- insert %s formula column to %s view" % (columnName, viewName)
    print "- formula %s" % (formula)
    viewObj = getViewObject(ixNet, viewName)
    formulaColumn = ixNet.add(viewObj + '/formulaCatalog', 'formulaColumn')
    ixNet.setAttribute(formulaColumn, '-caption', columnName)
    ixNet.setAttribute(formulaColumn, '-formula', formula)
    ixNet.commit()

################################################################################
# Define function to edit a formula column
################################################################################
def editFormulaColumn(ixNet, viewName, columnName, formula):
    print "- edit %s formula column %s in view" % (columnName, viewName)
    print "- new formula %s" % formula
    viewObj = getViewObject(ixNet, viewName)
    formulaColumns = ixNet.getList(viewObj + '/formulaCatalog', 'formulaColumn')
    for column in formulaColumns:
        if ixNet.getAttribute(column, '-caption') == columnName:
            ixNet.setAttribute(column, '-formula', formula)
            ixNet.commit()
            break

################################################################################
# Define function to compare 2 stats
################################################################################
def compareTwoStats(ixNet, viewName, statA, statB):
    print "- compare %s = %s" % (statA, statB)
    statsA = getValuesForStatInView(ixNet, viewName, statA)
    statsB = getValuesForStatInView(ixNet, viewName, statB)
    ipv4source = getValuesForStatInView(ixNet, viewName, "IPv4 :Source Address")
    for ip, st1, st2 in zip(ipv4source, statsA, statsB):
        if int(st1) == int(st2):
            print "\t- Source IP: %s --> OK " % ip
        else: 
            print "\t- Source IP: %s --> Failed: %s = %s, %s = %s " % (ip, statA, st1, statB, st2)

################################################################################
# Enable CSV Logging across all views 
################################################################################

print "Enable CSV Logging across all views"
setEnableCsvLogging(ixNet, True)

viewName = "Flow Statistics"
################################################################################
# Add Formula Column to view
################################################################################

print "Add Formula Column to view"
addFormulaColumn(ixNet, viewName, 'Formula Column Name', '="Tx Frames" * 2')

################################################################################
# Edit the Formula Column added to view
################################################################################

print "Edit the Formula Column added to view"
editFormulaColumn(ixNet, viewName, 'Formula Column Name', '="Tx Frames" * 3')

################################################################################
# Create a Snapshot CSV for view
################################################################################

print "Take a Snapshot CSV for view %s" % viewName
takeViewCSVSnapshot(ixNet, viewName)

################################################################################
# Check the Tx Frames = Rx Frames for each IPv4 source address
################################################################################

print "Check the Tx Frames = Rx Frames for each IPv4 source address"
compareTwoStats(ixNet, viewName, "Tx Frames", "Rx Frames")

################################################################################
# Disable CSV Logging across all views"
################################################################################

print "Disable CSV Logging across all views"
setEnableCsvLogging(ixNet, False)

File: LowLevelApi/NGPF/Perl/Statistics/Sample_Statistics_Operations.pl
################################################################################
# Version 1.0    $Revision: #1 $                                               #
#                                                                              #
#    Copyright  1997 - 2015 by IXIA                                            #
#    All Rights Reserved.                                                      #
#                                                                              #
#    Revision Log:                                                             #
#    02/04/2015 - Indranil Acharya - created sample                            #
#                                                                              #
################################################################################

################################################################################
#                                                                              #
#                                LEGAL  NOTICE:                                #
#                                ==============                                #
# The following code and documentation (hereinafter "the script") is an        #
# example script for demonstration purposes only.                              #
# The script is not a standard commercial product offered by Ixia and have     #
# been developed and is being provided for use only as indicated herein. The   #
# script [and all modifications enhancements and updates thereto (whether      #
# made by Ixia and/or by the user and/or by a third party)] shall at all times #
# remain the property of Ixia.                                                 #
#                                                                              #
# Ixia does not warrant (i) that the functions contained in the script will    #
# meet the users requirements or (ii) that the script will be without          #
# omissions or error-free.                                                     #
# THE SCRIPT IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND AND IXIA         #
# DISCLAIMS ALL WARRANTIES EXPRESS IMPLIED STATUTORY OR OTHERWISE              #
# INCLUDING BUT NOT LIMITED TO ANY WARRANTY OF MERCHANTABILITY AND FITNESS FOR #
# A PARTICULAR PURPOSE OR OF NON-INFRINGEMENT.                                 #
# THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SCRIPT  IS WITH THE #
# USER.                                                                        #
# IN NO EVENT SHALL IXIA BE LIABLE FOR ANY DAMAGES RESULTING FROM OR ARISING   #
# OUT OF THE USE OF OR THE INABILITY TO USE THE SCRIPT OR ANY PART THEREOF     #
# INCLUDING BUT NOT LIMITED TO ANY LOST PROFITS LOST BUSINESS LOST OR          #
# DAMAGED DATA OR SOFTWARE OR ANY INDIRECT INCIDENTAL PUNITIVE OR              #
# CONSEQUENTIAL DAMAGES EVEN IF IXIA HAS BEEN ADVISED OF THE POSSIBILITY OF    #
# SUCH DAMAGES IN ADVANCE.                                                     #
# Ixia will not be required to provide any software maintenance or support     #
# services of any kind (e.g. any error corrections) in connection with the     #
# script or any part thereof. The user acknowledges that although Ixia may     #
# from time to time and in its sole discretion provide maintenance or support  #
# services for the script any such services are subject to the warranty and    #
# damages limitations set forth herein and will not obligate Ixia to provide   #
# any additional maintenance or support services.                              #
#                                                                              #
################################################################################

################################################################################
#                                                                              #
# Description:                                                                 #
#    This script intends to demonstrate how to use NGPF Statistics Operation  API#
#    It will create SnapshotCSV and will publish it 			       #
#    than it will retrieve and display few statistics                          #
# Ixia Softwares:                                                              #
#    IxOS      6.80 EA (6.80.1100.7)                                           #
#    IxNetwork 7.50 EA (7.50.0.45)                                             #
#                                                                              #
################################################################################

################################################################################
# Please ensure that PERL5LIB environment variable is set properly so that 
# IxNetwork.pm module is available. IxNetwork.pm is generally available in
# C:\<IxNetwork Install Path>\API\Perl
################################################################################
use IxNetwork;
use strict;

sub assignPorts {
    my @my_resource = @_;
	my $ixNet    = $my_resource[0];
	my $chassis1 = $my_resource[1];
	my $card1    = $my_resource[2];
	my $port1    = $my_resource[3];
	my $chassis2 = $my_resource[4];
	my $card2    = $my_resource[5];
	my $port2    = $my_resource[6];
	my $vport1   = $my_resource[7];
	my $vport2   = $my_resource[8];
	
	my $root = $ixNet->getRoot();
	my $chassisObj1 = $ixNet->add($root.'/availableHardware', 'chassis');
    $ixNet->setAttribute($chassisObj1, '-hostname', $chassis1);
    $ixNet->commit();
    $chassisObj1 = ($ixNet->remapIds($chassisObj1))[0];
	
	my $chassisObj2 = '';
	if ($chassis1 ne $chassis2) {
	    $chassisObj2 = $ixNet->add($root.'/availableHardware', 'chassis');
        $ixNet->setAttribute($chassisObj2, '-hostname', $chassis2);
        $ixNet->commit();
        $chassisObj2 = ($ixNet->remapIds($chassisObj2))[0];
	} else {
	    $chassisObj2 = $chassisObj1;
	}
	
	my $cardPortRef1 = $chassisObj1.'/card:'.$card1.'/port:'.$port1;
    $ixNet->setMultiAttribute($vport1, '-connectedTo', $cardPortRef1,
        '-rxMode', 'captureAndMeasure', '-name', 'Ethernet - 001');
    $ixNet->commit();

    my $cardPortRef2 = $chassisObj2.'/card:'.$card2.'/port:'.$port2;
    $ixNet->setMultiAttribute($vport2, '-connectedTo', $cardPortRef2,
        '-rxMode', 'captureAndMeasure', '-name', 'Ethernet - 002');
		
    $ixNet->commit();
}

# Script Starts
print("!!! Test Script Starts !!!\n");

# Edit this variables values to match your setup
my $ixTclServer = '10.200.115.31';
my $ixTclPort   = '8009';
my @ports       = (('10.200.115.31', '5', '1'), ('10.200.115.31', '6', '1'));
# Spawn a new instance of IxNetwork object. 
my $ixNet = new IxNetwork();

print("Connect to IxNetwork Tcl server\n");
$ixNet->connect($ixTclServer, '-port', $ixTclPort, '-version', '7.50',
    '-setAttribute', 'strict');

print("Creating a new config\n");
$ixNet->execute('newConfig');

print("Adding 2 vports\n");
$ixNet->add($ixNet->getRoot(), 'vport');
$ixNet->add($ixNet->getRoot(), 'vport');
$ixNet->commit();

my @vPorts  = $ixNet->getList($ixNet->getRoot(), 'vport');
my $vportTx = $vPorts[0];
my $vportRx = $vPorts[1];
assignPorts($ixNet, @ports, $vportTx, $vportRx);
sleep(5);

print("Adding 2 topologies\n");
$ixNet->add($ixNet->getRoot(), 'topology', '-vports', $vportTx);
$ixNet->add($ixNet->getRoot(), 'topology', '-vports', $vportRx);
$ixNet->commit();

my @topologies = $ixNet->getList($ixNet->getRoot(), 'topology');
my $topo1 = $topologies[0];
my $topo2 = $topologies[1];

print("Adding 2 device groups\n");
$ixNet->add($topo1, 'deviceGroup');
$ixNet->add($topo2, 'deviceGroup');
$ixNet->commit();

my @t1devices = $ixNet->getList($topo1, 'deviceGroup');
my @t2devices = $ixNet->getList($topo2, 'deviceGroup');

my $t1dev1 = $t1devices[0];
my $t2dev1 = $t2devices[0];

print("Configuring the multipliers (number of sessions)\n");
$ixNet->setAttribute($t1dev1, '-multiplier', '1');
$ixNet->setAttribute($t2dev1, '-multiplier', '1');
$ixNet->commit();

print("Adding ethernet/mac endpoints\n");
$ixNet->add($t1dev1, 'ethernet');
$ixNet->add($t2dev1, 'ethernet');
$ixNet->commit();

my $mac1 = ($ixNet->getList($t1dev1, 'ethernet'))[0];
my $mac2 = ($ixNet->getList($t2dev1, 'ethernet'))[0];

print("Configuring the mac addresses\n");
$ixNet->setMultiAttribute($ixNet->getAttribute($mac1, '-mac').'/counter',
        '-direction', 'increment',
        '-start',     '18:03:73:C7:6C:B1',
        '-step',      '00:00:00:00:00:01');

$ixNet->setAttribute($ixNet->getAttribute($mac2, '-mac').'/singleValue',
        '-value', '18:03:73:C7:6C:01');
$ixNet->commit();

print("Add ipv4\n");
$ixNet->add($mac1, 'ipv4');
$ixNet->add($mac2, 'ipv4');
$ixNet->commit();

my $ip1 = ($ixNet->getList($mac1, 'ipv4'))[0];
my $ip2 = ($ixNet->getList($mac2, 'ipv4'))[0];

my $mvAdd1 = $ixNet->getAttribute($ip1, '-address');
my $mvAdd2 = $ixNet->getAttribute($ip2, '-address');
my $mvGw1  = $ixNet->getAttribute($ip1, '-gatewayIp');
my $mvGw2  = $ixNet->getAttribute($ip2, '-gatewayIp');

print("configuring ipv4 addresses\n");
$ixNet->setAttribute($mvAdd1.'/singleValue', '-value', '20.20.20.2');
$ixNet->setAttribute($mvAdd2.'/singleValue', '-value', '20.20.20.1');
$ixNet->setAttribute($mvGw1.'/singleValue', '-value', '20.20.20.1');
$ixNet->setAttribute($mvGw2.'/singleValue', '-value', "20.20.20.2");

$ixNet->setAttribute($ixNet->getAttribute($ip1, '-prefix').'/singleValue', '-value', '24');
$ixNet->setAttribute($ixNet->getAttribute($ip2, '-prefix').'/singleValue', '-value', '24');

$ixNet->setMultiAttribute($ixNet->getAttribute($ip1, '-resolveGateway').'/singleValue', '-value', 'true');
$ixNet->setMultiAttribute($ixNet->getAttribute($ip2, '-resolveGateway').'/singleValue', '-value', 'true');
$ixNet->commit();
sleep(5);

################################################################################
# Start protocol 
################################################################################

print("Starting protocol\n");
$ixNet->execute('startAllProtocols');
print("Running the protocol for 30 seconds ..\n");
sleep(30);

################################################################################
# Retrieve protocol statistics.
################################################################################

print("Fetching all Protocol Summary Stats\n");
my $viewPage = '::ixNet::OBJ-/statistics/view:"Protocols Summary"/page';
my @statcap  = $ixNet->getAttribute($viewPage, '-columnCaptions');
my @rowvals  = $ixNet->getAttribute($viewPage, '-rowValues');
my $index    = 0;
my $statValueList= '';
foreach $statValueList (@rowvals) {
    print("***************************************************\n");
    my $statVal = '';
    foreach $statVal (@$statValueList) {
	    my $statIndiv = ''; 
		$index = 0;
	    foreach $statIndiv (@$statVal) {
		    printf(" %-30s:%s\n", $statcap[$index], $statIndiv);
			$index++;
        }
    }    
}
print("***************************************************\n");

################################################################################
# Configure L2-L3 traffic 
################################################################################

print ("Congfiguring L2-L3 Traffic Item\n");
my $trafficItem1 = $ixNet->add(($ixNet->getRoot()).'/traffic', 'trafficItem');
$ixNet->setMultiAttribute($trafficItem1, '-name', 'Traffic Item 1',
    '-roundRobinPacketOrdering', 'false',
    '-trafficType', 'ipv4');
$ixNet->commit();

$trafficItem1    = ($ixNet->remapIds($trafficItem1))[0];
my $endpointSet1 = $ixNet->add($trafficItem1, 'endpointSet');
my @source       = ($topo2.'/deviceGroup:1/ethernet:1/ipv4:1');
my @destination  = ($topo1.'/deviceGroup:1/ethernet:1/ipv4:1');

$ixNet->setMultiAttribute($endpointSet1,
    '-name',                  'EndpointSet-1',
    '-multicastDestinations', [[]],
    '-scalableSources',       [],
    '-multicastReceivers',    [[]],
    '-scalableDestinations',  [],
    '-ngpfFilters',           [],
    '-trafficGroups',         [],
    '-sources',               @source,
    '-destinations',          @destination);
$ixNet->commit();

$ixNet->setMultiAttribute($trafficItem1.'/tracking',
    '-trackBy',        ['ipv4DestIp0','trackingenabled0'],
    '-fieldWidth',     'thirtyTwoBits',
    '-protocolOffset', 'Root.0',
    '-values',         []);
$ixNet->commit();

###############################################################################
# Apply and start L2/L3 traffic
###############################################################################

print("Applying L2/L3 traffic\n");
$ixNet->execute('apply', ($ixNet->getRoot()).'/traffic');
sleep(5);

print("Starting L2/L3 traffic\n");
$ixNet->execute('start', ($ixNet->getRoot()).'/traffic');
sleep(5);

print "#########################\n";
print "## Statistics Samples ##\n";
print "#########################\n";
print "";

################################################################################
# Define function to get the view object using the view name
################################################################################

sub getViewObject{
my @my_resource = @_;
	my $ixNet    = $my_resource[0];
	my $viewName = $my_resource[1];
    my @views = $ixNet->getList('/statistics', 'view');
    my $viewObj = '';
    my $editedViewName = '::ixNet::OBJ-/statistics/view:\"'.$viewName.'\"';
    foreach my $view (@views) {
        if ($editedViewName == $view) {
             $viewObj = $view;
            }
	}
    return $viewObj;
}

################################################################################
# Define function to get the values for the statistics in the view
################################################################################

sub getValuesForStatInView {
	my @my_resource = @_;
	my $ixNet    = $my_resource[0];
	my $statName = $my_resource[1];
	my $viewName = $my_resource[2];
    print "- get the stats for $statName in view $viewName";
    my @views = $ixNet->getList('/statistics', 'view');
    my $viewObj = getViewObject($ixNet, $viewName);
    my $returned_values = $ixNet->execute('getColumnValues', $viewObj, $statName);
    return $returned_values;
}

################################################################################
# Define function to get all the statistics in the view
################################################################################

sub getAllStatsInView {
	my @my_resource = @_;
	my $ixNet    = $my_resource[0];
	my $viewName = $my_resource[1];
    print "- get the stats in view $viewName";
    my $mview = getViewObject($ixNet, $viewName);
    my @mpage = $ixNet->getList($mview, 'page');
    my $mrowvalues = $ixNet->getAttribute(@mpage[0], '-rowValues');
    return $mrowvalues;
}

################################################################################
# Define function to create a Snapshot CSV
################################################################################

sub takeViewCSVSnapshot {
	my @my_resource = @_;
	my $ixNet    = $my_resource[0];
	my $viewName = $my_resource[1];
	my $csvPath="c:\\Regression\\Snapshot CSVs";
	my $csvType="currentPage";
    print "- take Snapshot CSV";
    my @SnapSettingList = [ 'Snapshot.View.Csv.Location: "'.$csvPath.'"',
                        'Snapshot.View.Csv.GeneratingMode: "kOverwriteCSVFile"',
                        'Snapshot.Settings.Name: '.$viewName, 
                        'Snapshot.View.Contents: '.$csvType ];
    $ixNet->execute('TakeViewCSVSnapshot','"'.$viewName.'"',@SnapSettingList);
    print "- snapshot CSV complete";
}

################################################################################
# Define function to Enable CSV Logging
################################################################################

sub setEnableCsvLogging {
	my @my_resource = @_;
	my $ixNet    = $my_resource[0];
	my $state='True';
    print "- set enableCsvLogging to: $state";
    $ixNet->setAttribute('/statistics', '-enableCsvLogging', $state);
    $ixNet->commit();
}

################################################################################
# Enable CSV Logging across all views 
################################################################################

print "Enable CSV Logging across all views";
setEnableCsvLogging($ixNet, 'True');
my $viewName = "Flow Statistics";

print "Take a Snapshot CSV for view $viewName";
takeViewCSVSnapshot($ixNet, $viewName);

################################################################################
# Stop protocol 
################################################################################

print("Stopping protocol\n");
$ixNet->execute('stopAllProtocols');

################################################################################
# Printing the data in the created Snapshot
################################################################################

print("Snapshot CSV file Information:\n");
my $filename = "c:\\Regression\\Snapshot CSVs\\Flow Statistics.csv";
open(my $fh, '<:encoding(UTF-8)', $filename)
  or die "Could not open file '$filename' $!";

while (my $row = <$fh>) {
   chomp $row;
   printf(join(" %s\n",split(/,/,$row)));
   print "\n";
   }
print "done\n";
print("!!! Test Script Ends !!!");